<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Detection of metacells from single-cell gene expression matrix — meta_cells • inferCSN</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Detection of metacells from single-cell gene expression matrix — meta_cells"><meta property="og:description" content="This function detects metacells from a single-cell gene expression matrix
using dimensionality reduction and clustering techniques."><meta property="og:image" content="https://mengxu98.github.io/inferCSN/logo.svg"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">inferCSN</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.8</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/mengxu98/inferCSN" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
<li>
  <a href="https://github.com/mengxu98" class="external-link">
    <span class="fas fa-home fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Detection of metacells from single-cell gene expression matrix</h1>
    <small class="dont-index">Source: <a href="https://github.com/mengxu98/inferCSN/blob/main/R/meta_cells.R" class="external-link"><code>R/meta_cells.R</code></a></small>
    <div class="hidden name"><code>meta_cells.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function detects metacells from a single-cell gene expression matrix
using dimensionality reduction and clustering techniques.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">meta_cells</span><span class="op">(</span></span>
<span>  <span class="va">matrix</span>,</span>
<span>  genes_use <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  genes_exclude <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  var_genes_num <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="fl">1000</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">matrix</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  gamma <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  knn_k <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  do_scale <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  pc_num <span class="op">=</span> <span class="fl">25</span>,</span>
<span>  fast_pca <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  do_approx <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  approx_num <span class="op">=</span> <span class="fl">20000</span>,</span>
<span>  directed <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  use_nn2 <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  cluster_method <span class="op">=</span> <span class="st">"walktrap"</span>,</span>
<span>  block_size <span class="op">=</span> <span class="fl">10000</span>,</span>
<span>  weights <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  do_median_norm <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-matrix">matrix<a class="anchor" aria-label="anchor" href="#arg-matrix"></a></dt>
<dd><p>Log-normalized gene expression matrix with rows as genes and columns as cells.</p></dd>


<dt id="arg-genes-use">genes_use<a class="anchor" aria-label="anchor" href="#arg-genes-use"></a></dt>
<dd><p>A vector of genes used to compute PCA, default is <em><code>NULL</code></em>.</p></dd>


<dt id="arg-genes-exclude">genes_exclude<a class="anchor" aria-label="anchor" href="#arg-genes-exclude"></a></dt>
<dd><p>A vector of genes to be excluded when computing PCA, default is <em><code>NULL</code></em>.</p></dd>


<dt id="arg-var-genes-num">var_genes_num<a class="anchor" aria-label="anchor" href="#arg-var-genes-num"></a></dt>
<dd><p>If <em><code>genes_use</code></em> is not provided, <em><code>var_genes_num</code></em> genes with the largest
variation are used. Default is <em><code>min(1000, nrow(matrix))</code></em>.</p></dd>


<dt id="arg-gamma">gamma<a class="anchor" aria-label="anchor" href="#arg-gamma"></a></dt>
<dd><p>Graining level of data (proportion of number of single cells in the initial dataset
to the number of metacells in the final dataset), default is <em><code>10</code></em>.</p></dd>


<dt id="arg-knn-k">knn_k<a class="anchor" aria-label="anchor" href="#arg-knn-k"></a></dt>
<dd><p>Parameter to compute single-cell kNN network, default is <em><code>5</code></em>.</p></dd>


<dt id="arg-do-scale">do_scale<a class="anchor" aria-label="anchor" href="#arg-do-scale"></a></dt>
<dd><p>Logical value, default is <em><code>TRUE</code></em>, whether to scale gene expression matrix
when computing PCA.</p></dd>


<dt id="arg-pc-num">pc_num<a class="anchor" aria-label="anchor" href="#arg-pc-num"></a></dt>
<dd><p>Number of principal components to use for construction of single-cell kNN network,
default is <em><code>25</code></em>.</p></dd>


<dt id="arg-fast-pca">fast_pca<a class="anchor" aria-label="anchor" href="#arg-fast-pca"></a></dt>
<dd><p>Logical value, default is <em><code>TRUE</code></em>, use <a href="https://rdrr.io/pkg/irlba/man/irlba.html" class="external-link">irlba</a> as a faster version
of prcomp.</p></dd>


<dt id="arg-do-approx">do_approx<a class="anchor" aria-label="anchor" href="#arg-do-approx"></a></dt>
<dd><p>Logical value, default is <em><code>FALSE</code></em>, compute approximate kNN in case of a large
dataset (&gt;50'000).</p></dd>


<dt id="arg-approx-num">approx_num<a class="anchor" aria-label="anchor" href="#arg-approx-num"></a></dt>
<dd><p>Number of cells to subsample for an approximate approach, default is <em><code>20000</code></em>.</p></dd>


<dt id="arg-directed">directed<a class="anchor" aria-label="anchor" href="#arg-directed"></a></dt>
<dd><p>Logical value, default is <em><code>FALSE</code></em>, whether to use directed graph.</p></dd>


<dt id="arg-use-nn-">use_nn2<a class="anchor" aria-label="anchor" href="#arg-use-nn-"></a></dt>
<dd><p>Logical value, default is <em><code>TRUE</code></em>, whether to use nn2 function.</p></dd>


<dt id="arg-seed">seed<a class="anchor" aria-label="anchor" href="#arg-seed"></a></dt>
<dd><p>Seed to use to subsample cells for an approximate approach, default is <em><code>1</code></em>.</p></dd>


<dt id="arg-cluster-method">cluster_method<a class="anchor" aria-label="anchor" href="#arg-cluster-method"></a></dt>
<dd><p>Clustering method to identify metacells, default is <em><code>walktrap</code></em>.
Available methods are <em><code>walktrap</code></em> and <em><code>louvain</code></em> (not recommended, gamma is ignored).</p></dd>


<dt id="arg-block-size">block_size<a class="anchor" aria-label="anchor" href="#arg-block-size"></a></dt>
<dd><p>Number of cells to map to the nearest metacell at a time (for approx
coarse-graining), default is <em><code>10000</code></em>.</p></dd>


<dt id="arg-weights">weights<a class="anchor" aria-label="anchor" href="#arg-weights"></a></dt>
<dd><p>Vector of cell weights (NULL by default),
used for computing average gene expression within cluster of metacells.</p></dd>


<dt id="arg-do-median-norm">do_median_norm<a class="anchor" aria-label="anchor" href="#arg-do-median-norm"></a></dt>
<dd><p>Logical value, default is <em><code>FALSE</code></em>,
whether to normalize by median value.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Additional parameters passed to other methods.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>A metacell matrix where rows represent metacells and columns represent genes.</p>
    </div>
    <div id="references">
    <h2>References</h2>
    <p>https://github.com/GfellerLab/SuperCell
https://github.com/kuijjerlab/SCORPION</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"example_matrix"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">meta_cells_matrix</span> <span class="op">&lt;-</span> <span class="fu">meta_cells</span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="va">example_matrix</span>,</span></span>
<span class="r-in"><span>  pc_num <span class="op">=</span> <span class="fl">10</span>,</span></span>
<span class="r-in"><span>  fast_pca <span class="op">=</span> <span class="cn">FALSE</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">meta_cells_matrix</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 500  18</span>
<span class="r-in"><span><span class="va">meta_cells_matrix</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6 x 6 Matrix of class "dgeMatrix"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>             g1        g10        g11        g12        g13        g14</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1,] 2.0545057 2.08367091 2.19895110 1.51677313 0.11024463 0.01935760</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [2,] 2.2880546 2.27451038 2.30015670 2.28168354 1.60320552 0.10528643</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [3,] 1.8903876 0.02312097 0.01589155 0.02682636 0.02432362 0.02568322</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [4,] 2.0238664 2.13578014 2.10612123 2.12087076 2.16751773 2.08957642</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [5,] 0.1361268 2.07312365 2.11331095 2.11336063 2.02598610 2.22708277</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [6,] 2.1689733 0.01482280 0.01606891 0.01890738 0.02267026 0.01781319</span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Meng Xu.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer></div>






  </body></html>

